src/udev/udevadm-lock.c

```diff
@@ -0,0 +1,402 @@
+                { "timeout", required_argument, NULL, 't'      },
+                { "print",   no_argument,       NULL, 'p'      },
+                {}
+        };
+
+        int c, r;
+
+        assert(argc >= 0);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, arg_print ? "hVd:b:t:p" : "+hVd:b:t:p", options, NULL)) >= 0)
+
+                switch (c) {
+
+                case 'h':
+                        return help();
+
+                case 'V':
+                        return print_version();
+
+                case 'd':
+                case 'b': {
+                        _cleanup_free_ char *s = NULL;
+                        char ***l = c == 'd' ? &arg_devices : &arg_backing;
+
+                        r = path_make_absolute_cwd(optarg, &s);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to make path '%s' absolute: %m", optarg);
+
+                        path_simplify(s);
+
+                        if (strv_consume(l, TAKE_PTR(s)) < 0)
+                                return log_oom();
+
+                        strv_uniq(*l);
+                        break;
+                }
+
+                case 't':
+                        r = parse_sec(optarg, &arg_timeout_usec);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to parse --timeout= parameter: %s", optarg);
+                        break;
+
+                case 'p':
+                        arg_print = true;
+                        break;
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        if (arg_print) {
+                if (optind != argc)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "No arguments expected");
+        } else {
+                if (optind + 1 > argc)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Too few arguments, command to execute.");
+
+                arg_cmdline = strv_copy(argv + optind);
+                if (!arg_cmdline)
+                        return log_oom();
+        }
+
+        if (strv_isempty(arg_devices) && strv_isempty(arg_backing))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "No devices to lock specified, refusing.");
+
+        return 1;
+}
+
+static int find_devno(
+                dev_t **devnos,
+                size_t *n_devnos,
+                const char *device,
+                bool backing) {
+
+        dev_t devt, whole_devt;
+        struct stat st;
+        int r;
+
+        assert(devnos);
+        assert(n_devnos);
+        assert(*devnos || *n_devnos == 0);
+        assert(device);
+
+        if (stat(device, &st) < 0)
+                return log_error_errno(errno, "Failed to stat '%s': %m", device);
+
+        if (S_ISBLK(st.st_mode))
+                devt = st.st_rdev;
```

Comment:
This means, `--backing=/dev/sda1` works well as `--device=/dev/sda1`, right?

P.S.
Ah, the man page explicitly say so. Then, why not to the --device accept both device node and regular file?

Reply:
I kept this separate for a reason, because I got burnt but an automatism like this myself when I developed systemd-repart. The thing is that if you use disk tools on regular files it could mean two things:

1. consider the regular file a disk image and operate on that as a regular file
2. find out the backing disk for that file, and operate on that

Tools like sfdisk/fdisk/cfdisk follow logic 1, while here we are tempted to follow logic 2. But if we pick the wrong logic the results can be fatal. And this happened to me as mentioned: by accident I repartitioned the disk my test image was on rather then the test image myself. And that sucked hard! :-(

I think it's much safer to assume people are looking for logic 1 instead of logic 2 when doing disk management and getting a path to a regular file passed. Hence I think logic 2 should be opt-in, i.e. explicitly require a concious decision to be used by users. My attempt with `--backing=` is exactly that: by naming hte option that way the user hopefully understands that this is about locking the *backing* disk, and we do not lock the regular file itself.

Does that make any sense?

Reply:
Indeed. I agree with you. Thanks.